### PR TITLE
Fix ze_peak explicit scaling benchmark

### PR DESCRIPTION
**Pull request by lyu**
The explicit scaling code for ze_peak violates L0 spec and has no overlap between sub-devices. This PR corrects these issues.

### CODE REVIEW

**Comment by lyu** [comment]
## Original execution flow for each subdevice ID
0. Assume current subdevice ID is N
1. Reset cmdlist N
2. Append 1 memcpy to cmdlist N
3. Close cmdlist N
4. For each warmup iteration: [submit cmdlist N to subdevice N, and if this is the last subdevice then synchronize the cmdqueue of **all subdevices**]
5. For each benchmark iteration: do the same as step 4
6. Measure and return the time taken to do step 5

Suppose we have two subdevices 0 & 1. For subdevice 0 there is no synchronization at all, since the cmdqueue is async we only measure the submission time which is very small. For subdevice 1 we will call cmdqueue sync on subdevice 0 at step 4, before we actually run the benchmark on subdevice 1 so there is no overlap at all.

At the end we sum all the time measurements and calculate the BW. Although we had no overlap, we also didn't measure execution on subdevice 0, so we get half of the actual time and thus double the BW, so this bug was not discovered before.

Additionally, for subdevice 0 we do submit-submit-...(500 times)...-submit-sync using the same cmdlist & cmdqueue pair, this violates L0 spec's description of `zeCommandQueueExecuteCommandLists` [ref](https://oneapi-src.github.io/level-zero-spec/level-zero/latest/core/api.html#zecommandqueueexecutecommandlists):

> The application must ensure the device is not currently referencing the command list since the implementation is allowed to modify the contents of the command list for submission.

So we saw command buffer GPU page faults when running ze_peak on PVC.

## Corrected execution flow for each subdevice ID
0. Assume current subdevice ID is N
1. If N == 0 then reset all cmdlists
2. Append 1 memcpy and 1 barrier to cmdlist N
3. Close cmdlist N
4. Run warmup iterations just on subdevice N
5. Reset cmdlist N
7. Append 500 [memcpy + barrier] to cmdlist N
8. Close cmdlist N
9. Submit cmdlist N to subdevice N, and if this is the last subdevice then synchronize the cmdqueue of **all subdevices**
10. Measure and return the time taken to do step 9

Basically, now we submit 1 cmdlist to each subdevice asynchronously, and we do synchronization on all subdevices once we have submitted all cmdlists. There will be some warmup & cmdlist operation overhead mixed in there, and the barriers also have their own overhead, but the measured BW is still very close to 2x the performance on a single subdevice.

File: perf_tests/ze_peak/src/ze_peak.cpp [hunk 1]
```diff
@@ -864,6 +867,12 @@ long double ZePeak::run_kernel(L0Context context, ze_kernel_handle_t &function,
         throw std::runtime_error("zeCommandListAppendLaunchKernel failed: " +
                                  std::to_string(result));
       }
+      result = zeCommandListAppendBarrier(
```
Comment (aravindksg) [inline perf_tests/ze_peak/src/ze_peak.cpp:870]: you could just do SUCCESS_OR_TERMINATE() ?

File: perf_tests/ze_peak/src/ze_peak.cpp [hunk 2]
```diff
@@ -894,35 +903,39 @@ long double ZePeak::run_kernel(L0Context context, ze_kernel_handle_t &function,
...
+        SUCCESS_OR_TERMINATE(zeCommandListAppendBarrier(
+            context.cmd_list[current_sub_device_id], nullptr, 0, nullptr));
       }
+      SUCCESS_OR_TERMINATE(
+          zeCommandListClose(context.cmd_list[current_sub_device_id]));
     }
 
     timer.start();
-    for (uint32_t i = 0; i < iters; i++) {
+    if (context.sub_device_count) {
```
Comment (aravindksg) [inline perf_tests/ze_peak/src/ze_peak.cpp:926]: so just to confirm - is the intention here that the following lines are executed
1. if we don't have flat device hierarchy and
2. if multiple sub-device are detected right?

might be good to add comments around here to clarify this, helps future code maintenance

File: perf_tests/ze_peak/src/ze_peak.cpp [hunk 3]
```diff
@@ -846,8 +846,11 @@ long double ZePeak::run_kernel(L0Context context, ze_kernel_handle_t &function,
 
   if (type == TimingMeasurement::BANDWIDTH) {
     if (context.sub_device_count) {
-      SUCCESS_OR_TERMINATE(
-          zeCommandListReset(context.cmd_list[current_sub_device_id]));
+      if (current_sub_device_id == 0) {
```
[the remark this answers is not in the document]
Comment (lyu) [reply perf_tests/ze_peak/src/ze_peak.cpp:849]: This marks the beginning of the benchmark, so we reset all cmdlists in the following loop. The previous version resets the cmdlists only when it's the subdevice's turn, and introduces a bit of overhead.

File: perf_tests/ze_peak/src/ze_peak.cpp (hunk 2 above)
Comment (lyu) [reply perf_tests/ze_peak/src/ze_peak.cpp:926]: Yeah, both conditions need to be hold for sub_device_count to be non-zero.
OK, I can add some comments to clarify the intention.

**Review by nrspruit — DISMISSED** [review]

File: perf_tests/ze_peak/src/ze_peak.cpp (hunk 2 above)
Comment (nrspruit) [reply perf_tests/ze_peak/src/ze_peak.cpp:928]: Missed this, can you add these comments?

Comment (lyu) [reply perf_tests/ze_peak/src/ze_peak.cpp:930]: I have added comments to clarify this but not on this specific line, since there're multiple occurrences of this condition. I have added one more comment in the first occurrence.

**Review by nrspruit — APPROVED** [review]

**Review by aravindksg — APPROVED** [review]
